precompiles/ArbWasmCache.go

```diff
@@ -0,0 +1,62 @@
+// Copyright 2024, Offchain Labs, Inc.
+// For license information, see https://github.com/nitro/blob/master/LICENSE
+
+package precompiles
+
+type ArbWasmCache struct {
+	Address addr // 0x72
+
+	UpdateProgramCache        func(ctx, mech, addr, bytes32, bool) error
+	UpdateProgramCacheGasCost func(addr, bytes32, bool) (uint64, error)
+}
+
+// See if the user is a cache manager owner.
+func (con ArbWasmCache) IsCacheManager(c ctx, _ mech, addr addr) (bool, error) {
+	return c.State.Programs().CacheManagers().IsMember(addr)
+}
+
+// Retrieve all authorized address managers.
+func (con ArbWasmCache) AllCacheManagers(c ctx, _ mech) ([]addr, error) {
+	return c.State.Programs().CacheManagers().AllMembers(65536)
+}
+
+// Caches all programs with the given codehash. Caller must be a cache manager or chain owner.
+func (con ArbWasmCache) CacheCodehash(c ctx, evm mech, codehash hash) error {
```

Comment:
There is no explicit connection between caching and eviction, and not limit here on number of total contracts hashed.
I understand why - but it makes the existence of multiple cache managers more risky than helpful IMHO. This level should only have one cacheManager (+possibly chain owner). If multiple contracts are needed - let them all request caching from the single cachemanager. Then the single cachemanager could e.g. give different sub-managers allowance or something..

Reply:
I like having multiple only in that it keeps our storage format the same in future ArbOS upgrades should we extend the affordances of these actors. Back when we were initially designing the chain owner system in Nitro and before that 3rd party aggregators (throwback) we had a similar convo.